File: src/reconciliation/MatchClassifier.cpp
#include "reconciliation/MatchClassifier.hpp"

#include <sstream>

namespace reconciliation::engine {

using models::Transaction;

std::string MatchClassifier::buildKey(const Transaction& t, MatchingStrategy strategy) {
    if (strategy == MatchingStrategy::ExactId) {
        return t.transactionId;
    }
    // CompositeKey: vendor_id + invoice_number + transaction_date, '|'-separated.
    // '|' is not a valid character in any of these three source fields
    // under this project's CSV format, so it's a safe, simple delimiter —
    // no escaping needed, unlike e.g. joining on a character that could
    // legitimately appear in a vendor ID.
    return t.vendorId + "|" + t.invoiceNumber + "|" + t.transactionDate.toString();
}

ReconciliationResult MatchClassifier::classifyMatch(const Transaction& internalTx,
                                                      const Transaction& externalTx,
                                                      const MatchingConfig& config) {
    ReconciliationResult result;
    result.transactionId = internalTx.transactionId;
    result.internalTransaction = internalTx;
    result.externalTransaction = externalTx;

    int64_t amountDiff = internalTx.amount.absDifferenceCents(externalTx.amount);
    int64_t taxDiff = internalTx.taxAmount.absDifferenceCents(externalTx.taxAmount);
    result.amountDifferenceCents = amountDiff;
    result.taxDifferenceCents = taxDiff;

    bool amountMismatch = amountDiff > config.amountToleranceCents;
    bool taxMismatch = taxDiff > config.amountToleranceCents;
    bool dateMismatch = internalTx.transactionDate != externalTx.transactionDate;
    bool vendorMismatch = internalTx.vendorId != externalTx.vendorId;

    int mismatchCount = (amountMismatch ? 1 : 0) + (taxMismatch ? 1 : 0) +
                         (dateMismatch ? 1 : 0) + (vendorMismatch ? 1 : 0);

    std::ostringstream details;
    if (amountMismatch) {
        details << "amount differs by " << amountDiff << " cents "
                << "(internal=" << internalTx.amount.toString()
                << ", external=" << externalTx.amount.toString() << "); ";
    }
    if (taxMismatch) {
        details << "tax differs by " << taxDiff << " cents "
                << "(internal=" << internalTx.taxAmount.toString()
                << ", external=" << externalTx.taxAmount.toString() << "); ";
    }
    if (dateMismatch) {
        details << "date differs (internal=" << internalTx.transactionDate.toString()
                << ", external=" << externalTx.transactionDate.toString() << "); ";
    }
    if (vendorMismatch) {
        details << "vendor differs (internal=" << internalTx.vendorId
                << ", external=" << externalTx.vendorId << "); ";
    }

    if (mismatchCount == 0) {
        result.status = ReconciliationStatus::Matched;
    } else if (mismatchCount > 1) {
        result.status = ReconciliationStatus::MultipleMismatches;
        result.mismatchDetails = details.str();
    } else if (amountMismatch) {
        result.status = ReconciliationStatus::AmountMismatch;
        result.mismatchDetails = details.str();
    } else if (taxMismatch) {
        result.status = ReconciliationStatus::TaxMismatch;
        result.mismatchDetails = details.str();
    } else if (dateMismatch) {
        result.status = ReconciliationStatus::DateMismatch;
        result.mismatchDetails = details.str();
    } else { // vendorMismatch
        result.status = ReconciliationStatus::VendorMismatch;
        result.mismatchDetails = details.str();
    }

    return result;
}

ReconciliationResult MatchClassifier::missingExternal(const Transaction& internalTx) {
    ReconciliationResult result;
    result.transactionId = internalTx.transactionId;
    result.internalTransaction = internalTx;
    result.status = ReconciliationStatus::MissingExternal;
    result.amountDifferenceCents = internalTx.amount.cents();
    result.mismatchDetails = "present in internal source only";
    return result;
}

ReconciliationResult MatchClassifier::missingInternal(const Transaction& externalTx) {
    ReconciliationResult result;
    result.transactionId = externalTx.transactionId;
    result.externalTransaction = externalTx;
    result.status = ReconciliationStatus::MissingInternal;
    result.amountDifferenceCents = externalTx.amount.cents();
    result.mismatchDetails = "present in external source only";
    return result;
}

ReconciliationResult MatchClassifier::duplicate(const Transaction& t) {
    ReconciliationResult result;
    result.transactionId = t.transactionId;
    if (t.source == models::SourceSystem::Internal) {
        result.internalTransaction = t;
    } else {
        result.externalTransaction = t;
    }
    result.status = ReconciliationStatus::Duplicate;
    result.mismatchDetails = "duplicate key within " +
        std::string(t.source == models::SourceSystem::Internal ? "internal" : "external") + " source";
    return result;
}

} // namespace reconciliation::engine

File: recon_engine/reconciliation/match_classifier.py
"""
MatchClassifier
----------------
Pure classification logic shared by all three reconciliation algorithms
(brute force, two-pointer, hash). This is what makes the three algorithms
*provably* consistent with each other: they differ only in HOW they find
candidate (internal, external) pairs / orphans / duplicate groups, never in
HOW a given pair is judged matched vs. mismatched vs. duplicate. Running the
same dataset through all three should always yield identical
matched/mismatched/duplicate counts — the algorithms only change the CPU
complexity of getting there.
"""
from __future__ import annotations

from ..models.transaction import SourceSystem, Transaction
from .config_and_result import MatchingConfig, MatchingStrategy, ReconciliationResult
from .status import ReconciliationStatus


def build_key(t: Transaction, strategy: MatchingStrategy) -> str:
    if strategy == MatchingStrategy.EXACT_ID:
        return t.transaction_id
    # COMPOSITE_KEY: vendor_id + invoice_number + transaction_date, '|'-separated.
    # '|' is not a valid character in any of these three source fields under
    # this project's CSV format, so it's a safe, simple delimiter — no
    # escaping needed, unlike e.g. joining on a character that could
    # legitimately appear in a vendor ID.
    return f"{t.vendor_id}|{t.invoice_number}|{t.transaction_date.to_string()}"


def classify_match(internal_tx: Transaction, external_tx: Transaction, config: MatchingConfig) -> ReconciliationResult:
    amount_diff = internal_tx.amount.abs_difference_cents(external_tx.amount)
    tax_diff = internal_tx.tax_amount.abs_difference_cents(external_tx.tax_amount)

    amount_mismatch = amount_diff > config.amount_tolerance_cents
    tax_mismatch = tax_diff > config.amount_tolerance_cents
    date_mismatch = internal_tx.transaction_date != external_tx.transaction_date
    vendor_mismatch = internal_tx.vendor_id != external_tx.vendor_id

    mismatch_count = sum([amount_mismatch, tax_mismatch, date_mismatch, vendor_mismatch])

    details_parts = []
    if amount_mismatch:
        details_parts.append(
            f"amount differs by {amount_diff} cents "
            f"(internal={internal_tx.amount.to_string()}, external={external_tx.amount.to_string()}); "
        )
    if tax_mismatch:
        details_parts.append(
            f"tax differs by {tax_diff} cents "
            f"(internal={internal_tx.tax_amount.to_string()}, external={external_tx.tax_amount.to_string()}); "
        )
    if date_mismatch:
        details_parts.append(
            f"date differs (internal={internal_tx.transaction_date.to_string()}, "
            f"external={external_tx.transaction_date.to_string()}); "
        )
    if vendor_mismatch:
        details_parts.append(
            f"vendor differs (internal={internal_tx.vendor_id}, external={external_tx.vendor_id}); "
        )
    details = "".join(details_parts)

    if mismatch_count == 0:
        status = ReconciliationStatus.MATCHED
        details = ""
    elif mismatch_count > 1:
        status = ReconciliationStatus.MULTIPLE_MISMATCHES
    elif amount_mismatch:
        status = ReconciliationStatus.AMOUNT_MISMATCH
    elif tax_mismatch:
        status = ReconciliationStatus.TAX_MISMATCH
    elif date_mismatch:
        status = ReconciliationStatus.DATE_MISMATCH
    else:  # vendor_mismatch
        status = ReconciliationStatus.VENDOR_MISMATCH

    return ReconciliationResult(
        transaction_id=internal_tx.transaction_id,
        internal_transaction=internal_tx,
        external_transaction=external_tx,
        status=status,
        amount_difference_cents=amount_diff,
        tax_difference_cents=tax_diff,
        mismatch_details=details,
    )


def missing_external(internal_tx: Transaction) -> ReconciliationResult:
    return ReconciliationResult(
        transaction_id=internal_tx.transaction_id,
        internal_transaction=internal_tx,
        status=ReconciliationStatus.MISSING_EXTERNAL,
        amount_difference_cents=internal_tx.amount.cents(),
        mismatch_details="present in internal source only",
    )


def missing_internal(external_tx: Transaction) -> ReconciliationResult:
    return ReconciliationResult(
        transaction_id=external_tx.transaction_id,
        external_transaction=external_tx,
        status=ReconciliationStatus.MISSING_INTERNAL,
        amount_difference_cents=external_tx.amount.cents(),
        mismatch_details="present in external source only",
    )


def duplicate(t: Transaction) -> ReconciliationResult:
    result = ReconciliationResult(
        transaction_id=t.transaction_id,
        status=ReconciliationStatus.DUPLICATE,
        mismatch_details=(
            "duplicate key within "
            + ("internal" if t.source == SourceSystem.INTERNAL else "external")
            + " source"
        ),
    )
    if t.source == SourceSystem.INTERNAL:
        result.internal_transaction = t
    else:
        result.external_transaction = t
    return result
